File: src/contest/ccc/CCC_2016_S3.java
package contest.ccc;

import java.util.*;
import java.io.*;

public class CCC_2016_S3 {

	static BufferedReader br;
	static PrintWriter out;
	static StringTokenizer st;

	static int N, M;
	static ArrayList<ArrayList<Integer>> adj = new ArrayList<ArrayList<Integer>>();
	static boolean[] isPho;
	static boolean[] marked;
	static int[] dist;
	static int totalDist;

	public static void main (String[] args) throws IOException {
		br = new BufferedReader(new InputStreamReader(System.in));
		out = new PrintWriter(new OutputStreamWriter(System.out));
		//br = new BufferedReader(new FileReader("in.txt"));
		//out = new PrintWriter(new FileWriter("out.txt"));

		N = readInt();
		M = readInt();

		isPho = new boolean[N];
		marked = new boolean[N];

		for (int i = 0; i < N; i++)
			adj.add(new ArrayList<Integer>());

		for (int i = 0; i < M; i++)
			isPho[readInt()] = true;

		for (int i = 0; i < N - 1; i++) {
			int a = readInt();
			int b = readInt();
			adj.get(a).add(b);
			adj.get(b).add(a);
		}

		for (int i = 0; i < N; i++)
			if (isPho[i]) {
				mark(i, -1);
				break;
			}

		dist = new int[N];
		for (int i = 0; i < N; i++)
			if (marked[i]) {
				computeDist(i, -1, 0);
				break;
			}

		int maxIndex = -1;
		for (int i = 0; i < N; i++)
			if (marked[i] && (maxIndex == -1 || (dist[maxIndex] < dist[i])))
				maxIndex = i;

		dist = new int[N];
		computeDist(maxIndex, -1, 0);

		int longestDist = 0;

		for (int i = 0; i < N; i++)
			longestDist = Math.max(longestDist, dist[i]);

		out.println(totalDist - longestDist);
		out.close();
	}

	static void computeDist (int u, int prev, int depth) {
		dist[u] = depth;
		for (int v : adj.get(u)) {
			if (v != prev && marked[v]) {
				computeDist(v, u, depth + 1);
			}
		}
	}

	static void mark (int u, int prev) {
		if (isPho[u])
			marked[u] = true;

		for (int v : adj.get(u)) {
			if (v == prev)
				continue;
			mark(v, u);
			if (marked[v]) {
				totalDist += 2;
				marked[u] = true;
			}
		}
	}

	static String next () throws IOException {
		while (st == null || !st.hasMoreTokens())
			st = new StringTokenizer(br.readLine().trim());
		return st.nextToken();
	}

	static long readLong () throws IOException {
		return Long.parseLong(next());
	}

	static int readInt () throws IOException {
		return Integer.parseInt(next());
	}

	static double readDouble () throws IOException {
		return Double.parseDouble(next());
	}

	static char readCharacter () throws IOException {
		return next().charAt(0);
	}

	static String readLine () throws IOException {
		return br.readLine().trim();
	}
}

File: src/contest/ccc/CCC_2016_S3.cpp
#include <bits/stdc++.h>

#define mp make_pair
#define pb push_back
#define INF 1<<30
#define MOD 1000000007
#define rint(x) scanf("%d", &(x))
#define rlong(x) scanf("%lld", &(x))
#define SIZE 100100
#define l(x) x << 1
#define r(x) x << 1|1
#define m(x, y) (x + y)/2

using namespace std;

typedef long long ll;
typedef pair<int, int> pi;
typedef pair<ll, ll> pll;

int N, M;
vector<vector<int>> adj (SIZE);
bool isPho[SIZE], marked[SIZE];
int dist[100000];
int totalDist = 0;

void computeDist (int u, int prev, int depth) {
    dist[u] = depth;
    for (int v : adj[u]) {
        if (v != prev && marked[v]) {
            computeDist(v, u, depth + 1);
        }
    }
}

void mark (int u, int prev) {
    if (isPho[u])
        marked[u] = true;
    for (int v : adj[u]) {
        if (v == prev)
            continue;
        mark(v, u);
        if (marked[v]) {
            marked[u] = true;
            totalDist += 2;
        }
    }
}

int main () {
	// freopen("in.txt", "r", stdin);
	// freopen("out.txt", "w", stdout);

    rint(N);
    rint(M);

    for (int i = 0; i < M; i++) {
        int x;
        rint(x);
        isPho[x] = true;
    }

    for (int i = 0; i < N - 1; i++) {
        int a, b;
        rint(a), rint(b);
        adj[a].pb(b);
        adj[b].pb(a);
    }
    for (int i = 0; i < N; i++) {
        if (isPho[i]) {
            mark(i, -1);
            break;
        }
    }

    memset(dist, 0, sizeof dist);

    for (int i = 0; i < N; i++) {
        if (marked[i]) {
            computeDist(i, -1, 0);
            break;
        }
    }

    int maxIndex = -1;
    for (int i = 0; i < N; i++)
        if (marked[i] && (maxIndex == -1 || (dist[maxIndex] < dist[i])))
            maxIndex = i;

    memset(dist, 0, sizeof dist);
    computeDist(maxIndex, -1, 0);

    int longestDist = 0;

    for (int i = 0; i < N; i++)
        longestDist = max(longestDist, dist[i]);

    printf("%d\n", totalDist - longestDist);

	return 0;
}
